google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadChannel.java

```diff
@@ -101,27 +101,28 @@ public int read(ByteBuffer dst) throws IOException {
       checkOpen();
       int amt;
       int retries = 0;
-      int maxRetries = 3;
+      int maxRetries = Math.max(3, maxChannelReopens);
       dst.mark();
       while (true) {
         try {
           dst.reset();
           amt = channel.read(dst);
           break;
         } catch (StorageException exs) {
-          if (exs.getMessage().contains("Connection closed prematurely") && reopens < maxChannelReopens) {
-            // this error isn't marked as retryable since the channel is closed;
-            // but here at this higher level we can retry it.
+          if (reopens < maxChannelReopens && (isReopenable(exs) || isReopenable(exs.getCause()))) {
```

Comment:
Reiterate my recommendation to iterate all the way down the chain of `getCause()` to the bottom, as it's not clear to me at all that the exceptions `isReopenable()` checks for will be at or near the top of the pile.